clients/githubrepo/search.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package githubrepo
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-github/v38/github"
+
+	"github.com/ossf/scorecard/v2/clients"
+)
+
+type searchHandler struct {
+	ghClient *github.Client
+	ctx      context.Context
+	owner    string
+	repo     string
+}
+
+func (handler *searchHandler) init(ctx context.Context, owner, repo string) error {
+	handler.ctx = ctx
+	handler.owner = owner
+	handler.repo = repo
+	return nil
+}
+
+func (handler *searchHandler) search(request clients.SearchRequest) (clients.SearchResponse, error) {
+	query, err := handler.buildQuery(request)
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("%w", err)
+	}
+
+	resp, _, err := handler.ghClient.Search.Code(handler.ctx, query, &github.SearchOptions{})
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("Search.Code: %w", err)
+	}
+	return searchResponseFrom(resp), nil
+}
+
+func (handler *searchHandler) buildQuery(request clients.SearchRequest) (string, error) {
+	if request.Query == "" {
+		// nolint: goerr113
+		return "", errors.New("search query is empty")
+	}
+	var queryBuilder strings.Builder
+	if _, err := queryBuilder.WriteString(
```

Comment:
Why do three api calls to queryBuilder.WriteString if the request.Filename and request.Path are set. so i think better to build the string first and then do just one queryBuilder.WriteString call.

Reply:
Not sure how you mean `build the string first`. `queryBuilder.WriteString` is just an optimized way of concatenating strings in Golang.

Reply:
+1 with @azeemshaikh38 on this.

Reply:
Ok, thanks for explaining, i got confused if this was making an actual github api call.